pkg/credentialprovider/apis/dockerauth/types.go

```diff
@@ -0,0 +1,99 @@
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type DockerAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string `json:"name"`
+
+	// imageMatchers is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// DockerAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each ImageMatcher string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	ImageMatchers []string `json:"imageMatchers"`
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg `json:"extraArgs,omitempty"`
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+type PluginArg struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginRequest is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
+type DockerAuthPluginRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// image is the container image that is being pulled as part of the
+	// container registry credential request.
+	Image string `json:"image"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginResponse holds credentials for the kubelet runtime
+// to use for image pulls.  It is returned from the plugin as stdout.
+type DockerAuthPluginResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// username is the username used for authenticating to the Docker registry
+	// +optional
+	Username *string `json:"username,omitempty"`
+	// password is the password used for authenticating to the Docker registry
```

Comment:
I wonder if we should plan ahead here. It seems like we will want a union  of credential options. Might be good to put a enumeration of auth types now (None, username/password) and it will be easier later to extend and clearly understand the desired behavior. We can add auth types like one-time-token, client cert, and have it clear whats happening.

Reply:
In general I agree we should support multiple auth types, but the keyring today _only_ consumes username and password from the existing providers:

https://github.com/kubernetes/kubernetes/blob/54df1fdc9083ecc459ba60f395d755a90a2cd0b8/pkg/credentialprovider/keyring.go#L83-L87

I feel like that should be updated before we add more things to this API to support other auth types.

Reply:
Agree with Andrew here, the limiting factor is the keyring... and there's nothing stopping us from updating this response object in the next version since we have included `TypeMeta`, right?

Reply:
+1 for externalizing the existing format